feat: add k2 compiler migration guide (#4137)
JetBrains/kotlin-web-site: docs/topics/k2-compiler-guide.md

in module one:

```kotlin
// Module three
fun takeString(g: Generic<Some<String>>) {}

fun test() {
    // Triggers a warning in Kotlin 2.0.0
    val z = gen()

    // Triggers an error in Kotlin 2.0.0
    takeString(z)
}
```

**What's the best practice now?**

To avoid these problems, you can configure a direct dependency for module three on module one.

Alternatively, you can adapt your code to make the types accessible within the same module.

For more information, see the [corresponding issue in YouTrack](https://youtrack.jetbrains.com/issue/KT-64474).

### Consistent resolution order of Kotlin properties and Java fields with the same name

**What's changed?**

Before Kotlin 2.0.0, if you worked with Java and Kotlin classes that inherited from each other and contained Kotlin
properties and Java fields with the same name, the resolution behavior of the duplicated name was inconsistent. There was
also conflicting behavior between IntelliJ IDEA and the compiler. For the latest version of Kotlin, when developing
the new resolution behavior we aimed to cause the least impact to users.

> For example, suppose there was a Java class `Base`:

Reviewer: ```suggestion
For example, suppose there is a Java class `Base`:
```

```java
public class Base {
    public String a = "a";

    public String b = "b";
}
```

> Let's say there was also a Kotlin class `Derived` that inherits from the aforementioned `Base` class:

Reviewer: ```suggestion
Let's say there is also a Kotlin class `Derived` that inherits from the aforementioned `Base` class:
```

```kotlin
class Derived : Base() {
    val a = "aa"

>     // Declare custom get() function

Reviewer: ```suggestion
    // Declares custom get() function
```

    val b get() = "bb"
}

fun main() {
    // Resolves Derived.a
    println(a)
    // aa

    // Resolves Base.b
    println(b)
    // b
}
```

> Given the above, `a` was resolved to the Kotlin property within the `Derived` Kotlin class, whereas `b` was resolved to

Reviewer: ```suggestion
Given the above, `a` resolves to the Kotlin property within the `Derived` Kotlin class, whereas `b` resolves to
```

### Improved null safety for Java primitive arrays

**What's changed?**

Starting with Kotlin 2.0.0, the compiler correctly infers the nullability of Java primitive arrays imported to Kotlin.
Now, it retains native nullability from the `TYPE_USE` annotations used with Java primitive arrays and emits errors when
their values are not used according to annotations.

Usually, when Java types with `@Nullable` and `@NotNull` annotations are called from Kotlin, they receive the appropriate
native nullability:

```java
interface DataService {
    @NotNull ResultContainer<@Nullable String> fetchData();
}
```
```kotlin
val dataService: DataService = ... 
dataService.fetchData() // -> ResultContainer<String?>
```

Previously, however, when Java primitive arrays were imported to Kotlin, all `TYPE_USE` annotations were lost, resulting
in platform nullability and possibly unsafe code:

```java
interface DataProvider {
    int @Nullable [] fetchData();
}
```

```kotlin
val dataService: DataProvider = ...
dataService.fetchData() // -> IntArray .. IntArray?
// No error, even though `dataService.fetchData()` might be `null` according to annotations

> // This might result in the NullPointerException

Reviewer: ```suggestion
// This might result in a NullPointerException
```

dataService.fetchData()[0]
```
Note that this issue never affected nullability annotations on the declaration itself, only the `TYPE_USE` ones.

**What's the best practice now?**

In Kotlin 2.0.0, null safety for Java primitive arrays is now standard in Kotlin, so check your code for new warnings
and errors if you use them:

* Any code that uses a `@Nullable` Java primitive array without an explicit nullability check or attempts to pass `null`
  to a Java method expecting a non-nullable primitive array will now fail to compile.
* Using a `@NotNull` primitive array with a nullability check now emits "Unnecessary safe call" or "Comparison with null
  always false" warnings.

For more information, see the [corresponding issue in YouTrack](https://youtrack.jetbrains.com/issue/KT-54521).

### Per subject area

These subject areas list changes that are unlikely to affect your code but provide links to the relevant YouTrack issues

> for further reading. Changes that are listed with an asterisk (*) are explained at the beginning of the section.

Reviewer: ```suggestion
for further reading. Changes listed with an asterisk (*) next to the Issue ID are explained at the beginning of the section.
```
